mlir/include/mlir/IR/OpDefinition.h

```diff
@@ -1593,7 +1593,8 @@ foldTrait(Operation *, ArrayRef<Attribute>, SmallVectorImpl<OpFoldResult> &) {
 /// Given a tuple type containing a set of traits, return the result of folding
 /// the given operation.
 template <typename... Ts>
-static LogicalResult foldTraits(Operation *op, ArrayRef<Attribute> operands,
+static LogicalResult foldTraits([[maybe_unused]] Operation *op,
+                                [[maybe_unused]] ArrayRef<Attribute> operands,
```

Comment:
This and similar seem clearly used unless `sizeof...(Ts)` is 0. Would adding a static assert potentially fix that?
Otherwise the negative of declaring these as maybe unused has the downside that if they truly become unused at some point in the future,  we won't get that warning

Reply:
`static_assert(sizeof(op) && sizeof(operands)); // avoid unused warning`
also avoids the compiler message.
Would that be acceptable?

Reply:
Well, this would probably not address your concern, though.

Reply:
I could offer the following, which might be a bit too much for the gain:
```C++
template <typename... Ts, std::size_t _i = sizeof...(Ts), std::enable_if_t<(_i > 0), int> = 0>
static LogicalResult foldTraits(Operation *op,
                                ArrayRef<Attribute> operands,
                                SmallVectorImpl<OpFoldResult> &results) {
  return success((succeeded(foldTrait<Ts>(op, operands, results)) || ...));
}
// this specialization avoid compiler warnings about unused params
template <typename... Ts, std::size_t _i = sizeof...(Ts), std::enable_if_t<(_i == 0), int> = 0>
static LogicalResult foldTraits([[maybe_unused]] Operation *op,
                                [[maybe_unused]] ArrayRef<Attribute> operands,
                                SmallVectorImpl<OpFoldResult> &results) {
  return success((succeeded(foldTrait<Ts>(op, operands, results)) || ...));
}
```

Reply:
another workaround is to add a base case and then the rest of the parameter pack:
```c++
template <typename First, typename... Rest>
static LogicalResult foldTraits([[maybe_unused]] Operation *op,
                                [[maybe_unused]] ArrayRef<Attribute> operands,
                                SmallVectorImpl<OpFoldResult> &results) {
  return success((succeeded(foldTrait<First>(op, operands, results) || succeeded(foldTrait<Rest>(op, operands, results)) || ...));
}                  
```

Reply:
but to me this seems like an overzealous warning and any the workarounds only make the code less readable...

Reply:
I suggest to just leave the `[[maybe_unused]]`, in particular since the warning does not appear when I add an actually unused parameter.

Reply:
@dwblaikie what do you think about this?

Reply:
I believe we already disable this warning (is it this one?) with some known version of GCC that have false positives: https://github.com/llvm/llvm-project/blob/main/mlir/CMakeLists.txt#L87-L94

Reply:
Yes, it's the same warning (`-Wunused-but-set-parameter`) as handled in the CMakeLists.txt that you referred to. The warning triggers for gcc 11, 12, and 13 (didn't try other versions).

What's so bad about using `maybe_unused`? It works around single instance (in a very small function) of a questionable trigger of a warning. In my mind, disabling a warning for the entire project potentially does more harm.

Sure, I can disable the warning if that's preferred. I can simply adjust the version check. Let me know.

Reply:
Ok, I disabled the warning on gcc<14 and removed my previously added [[maybe_unused]] in related places.